### DISCUSSION TITLE
led-strip

### DISCUSSION THREAD

**Discussion by illidan112** [UncleRus/esp-idf-lib#366] [Q&A] [answered]
Hi!
I created project with Wi-fi in sta_mode and with your library for ws2812. This project: https://github.com/illidan112/wireless-led .
Everything works, but unfortunately the strip sometimes flashes with unnecessary colors for a split second, but this is quite noticeable to the eye. I saw warning table on documentary page, so wifi opening on core=
![led](https://user-images.githubusercontent.com/50042793/175024694-40b83610-33fb-4a3e-ad10-d5497ce8e997.png)
![wifi](https://user-images.githubusercontent.com/50042793/175024698-d12e6156-8351-4301-b755-45f39b5ec197.png)
0 and my "xRainbowFade" task on core=1, but this didn't help. Can you give me advise how can i fix it?

P.s. I'm sure that the reason for the bug is wi-fi, because if you do not start it, then everything works fine.

**Comment by UncleRus** [comment, marked as answer]
It looks like you can't, at least as long led_strip uses legacy rmt driver.
I ran into the same failures in my own project: https://github.com/UncleRus/EvLamp .
The main reason for such color bursts is sudden pauses in the RMT transmission, which is clearly visible on the oscilloscope.
WS2812B, of course, treats such pauses as valid bits and, as a result, all LEDs in the chain set completely random colors after bad timing.
In my project, such failures are not a big problem, since wifi is used only for the API, but in your case it seems to be a blocking bug.

What to try:
1. Be sure to call the `led_strip_init()` function from a task that is **already** running on the second core. Calling `led_strip_init()` from a task running on the first core and then calling `led_strip_flush()` from a task running on the second core will not work. **All** work with the led_strip should be done on the second core.
2. Try to rewrite led_strip to use new RMT drivers (see #331). Not sure if this will help.
3. Increase the size of internal RMT buffers.
4. Discard the RMT driver from ESP-IDF and write it completely on your own, working directly with the hardware, DMA and interrupt registers.

**Reply by illidan112** [reply]
The reason was in the first paragraph. I called led_strip_init() without freertos, therefore it was called on 0 core, like wifi. I changed core for wifi to 1 and the bug was gone. Can I ask some clarifying questions in Russian here?

**Reply by UncleRus** [reply]
>  Can I ask some clarifying questions in Russian here?

Конечно

**Reply by illidan112** [reply]
Про поменять RMT драйвер на новый. Не совсем понимаю о чём речь?

**Reply by UncleRus** [reply]
Реализация драйвера RMT, который используется в led_strip, в последнем мастере ESP-IDF (>= v5.0.0) помечена устаревшей. При сборке выдается соответствующий ворнинг. Вместо нее предлагается использовать новый совершенно упоротый и, КМК, неудобный драйвер.
